### DISCUSSION TITLE
Is it possible to incorporate the rules of two or more grammars?

### DISCUSSION THREAD

**Discussion by brunocartaxo** [tree-sitter/tree-sitter#3805] [Q&A] [answered]
### Problem

As far as I could see, the grammar function can receive just one argument as baseGrammar. I would like to know if it is possible to make my grammar incorporate the rules of two or more grammars.

I could not find this information in the tree-sitter documentation.

**Comment by ObserverOfTime** [comment]
No but you can probably do something like this:

```javascript
const A = require('tree-sitter-a/grammar');
const B = require('tree-sitter-b/grammar');

module.exports = grammar(A, {
  rules: {
    ...B.grammar.rules
  }
});
```

**Comment by brunocartaxo** [comment]
thank you for answering me @ObserverOfTime 

I tried what you recommended `...js_grammar.grammar.rules`, but I obtained this error:

```
...js_grammar.grammar.rules,
                              ^

TypeError: Cannot read properties of undefined (reading 'rules')
```

So, I tried `...js_grammar.rules` instead, and obtained this other error:

```
Error: Grammar rules must all be functions. 'program' rule is not.
```

following is the grammar I'm trying to generate a parser for:

```
const js_grammar = require('tree-sitter-javascript/grammar');
const my_base_grammar = require('tree-sitter-my-base/grammar');

module.exports = grammar(my_base_grammar, {
    name: 'my_js_grammar', 

    rules: {

        ...js_grammar.rules,
    }
});
```

the `js_grammar` is the default tree-sitter javascript grammar. It indeed has a rule named `program`, however it seems that tree sitter is not recognizing the rule as a function.

**Comment by ObserverOfTime** [comment]
If you can't just extend JS in your "base" grammar, you can try injections instead.

**Reply by brunocartaxo** [reply]
unfortunately, I cannot make my base grammar extend js because i use my base grammar with other languages. the idea is modularize the base grammar to reuse with many languages (javascript is one of them).

what do you mean " try injections"?

**Comment by amaanq** [comment]
You can't spread another grammar's rules since the `grammar` function has already been run on them, so they're effectively json-like objects you can reuse in a rule's definition but not as a rule itself. This is not really well-defined, but if you take a look at `dsl.js` and/or `dsl.d.ts` you can see the expected object type that `js_grammar.grammar.rules` looks like.

**Reply by brunocartaxo** [reply]
thank you for your reply @amaanq 

i took a look at the dsl files you mentioned, and as far as i could understand, it defines the types of the rules and the grammar function implementation. however, i could not guess if it is possible to make my grammar extend more than one grammar. do you have any suggestion?

**Reply by amaanq** [reply]
Oops, sorry I missed this. Anyway, glad you figured it out :slightly_smiling_face:

**Comment by brunocartaxo** [comment, marked as answer]
I have solved the problem creating the base grammar object, but not calling the grammar function.

So, my base grammar file (tree-sitter-my-base/grammar.js) is something like this:

```javascript
module.exports = { // NOT CALLING THE GRAMMAR FUNCTION PASSING THIS GRAMMAR OBJECT
    rules: {
        // MY RULES
    }
}
```

And my custom javascript grammar is something like this:

```javascript
const js_grammar = require('tree-sitter-javascript/grammar');
const my_base_grammar = require('tree-sitter-my-base/grammar');

module.exports = grammar(js_grammar, {
    name: 'my_js_grammar', 

    rules: {
        ...my_base_grammar.rules,
        
         // MY CUSTOM JAVASCRIPT RULES
    }
});
```
